### ISSUE TITLE
Transaction detail screen shows transaction time inconsistently – UTC in header, localised/merchant timezone in timeline

### ISSUE THREAD

**Issue by haszari** [Automattic/woocommerce-payments#9135]
### Describe the bug
The transaction detail screen shows transaction summary and a timeline of various events.

- The summary shows `DATE` in UTC.
- The timeline shows events in local timezone (not sure if it's browser or woo-store setting).

### To Reproduce
1. Be a shopper or merchant in a non-UTC timezone.
2. Make a purchase.
3. View the transaction details for the relevant charge.
4. See different times on same page.

### Screenshots

<img width="1223" alt="Screenshot 2024-07-19 at 11 44 07 AM" src="https://github.com/user-attachments/assets/92eff706-5b49-4703-adcf-54c107866af3">

Note I am in NZT, UTC+12.

### Expected behavior
Ideally, all times in merchant dashboard should use either merchant/store timezone, or user (browser) current timezone. 

Note that this is likely inconsistent in a few places in UI, so this might be a symptom of a larger problem. Many screens use UTC (in particular, list views – transactions, disputes etc).

### Additional context
Discovered when testing / working on `Payment overview > Payment activity card` project, see related issue:

- https://github.com/Automattic/woocommerce-payments/issues/9133
- https://github.com/Automattic/woocommerce-payments-server/issues/5858

**Comment by souravdebnath1986** [comment]
Thanks for catching this! 

> all times in merchant dashboard should use either merchant/store timezone, or user (browser) current timezone

All times should be in the **merchant/store timezone**. This makes analytics simpler as there could be a desire to analyse a specific period's activity of the merchant store by merchant users who might not be in the same location as the store. 

Are their other inconsistencies within our Payments UX in wcadmin that we know of ? For example, is the total balance and available balance shown in the Payments Overview grouped in UTC or merchant/store or browser timezone (my guess is merchant/store if this value is directly read from Stripe) ?

**Comment by haszari** [comment]
> Are their other inconsistencies within our Payments UX in wcadmin that we know of ? 

I don't know the full picture, IMO we need to do at least a high-level audit, I suspect there are more places affected than this. 

There's another issue open for UTC/timezone inconsistencies with CSV exports:

- https://github.com/Automattic/woocommerce-payments/issues/8522

That's all I could find logged, might be more if we test / review systematically.

**Comment by haszari** [comment]
> IMO we need to do at least a high-level audit, I suspect there are more places affected than this.

I did some more research while [reviewing](https://github.com/Automattic/woocommerce-payments/pull/9461#pullrequestreview-2323875639) Automattic/woocommerce-payments#9461 and found a bunch of similar issues (see below). I think we need to audit all dates/times and confirm / refine design so we can fix all and ensure things are consistent. 

I don't know if it's safe to try to fix each instance in isolation (at least I don't want to assume that's safe!).

##### open issues
- https://github.com/Automattic/woocommerce-payments/issues/8522
- https://github.com/Automattic/woocommerce-payments-server/issues/2847
- https://github.com/Automattic/woocommerce-payments-server/issues/4845
- https://github.com/Automattic/woocommerce-payments/issues/9472
- https://github.com/Automattic/woocommerce-payments/issues/9135 - timeline view, `focus: payments acceptance & processing` - not `reporting`

#### merged PRs / closed issues
- https://github.com/Automattic/woocommerce-payments/pull/9162
- https://github.com/Automattic/woocommerce-payments/pull/4711
- https://github.com/Automattic/woocommerce-payments-server/issues/2275

FYI @deepakpathania if we make progress on the above issues, we might pick up this one as well so we can tidy it up in one go.

**Comment by haszari** [comment]
Marking as `blocked` until we have audited and aligned on a path forward:

- Automattic/woocommerce-payments-server#6646

**Comment by haszari** [comment]
Re-labelling this as `reporting`, since it's reporting related for Helix and best counted as part of "reporting" bucket.

**Comment by Jinksi** [comment]
After investigating a solution to this, the cause and best place to fix this issue is with the [`@woocommerce/components` `Timeline` component](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/timeline/timeline-item.js#L11), assuming we want this view to show datetimes in the store's TZ:

It uses `@wordpress/date` `format()` to convert the provided date to a string ([src](https://github.com/woocommerce/woocommerce/blob/bccfa97183edb35919c8f092f1fffa0bb05fd3a7/packages/js/components/src/timeline/timeline-item.js#L11)), which will apply the browser's local timezone offset to the rendered time format.

If instead `Timeline` used `@wordpress/date` `datei18n()` to use the site/store's WP Settings timezone offset, the rendered datetimes would match the rendered time in the header.


```diff
-const itemTimeString = format( clockFormat, item.date );
+const itemTimeString = datei18n( clockFormat, item.date );
```


Example of how a UTC datetime of `1 January 2025 at 11:11` is converted by each function on my test store (store TZ`UTC-5`, browser TZ `UTC+10`):
```js
testDate = '2025-01-01T11:11:11.000Z'
dateI18n( 'c', testDate ) = '2025-01-01T06:11:11-05:00' // converts to the store's timezone of `UTC-5`
format( 'c', testDate )   = '2025-01-01T21:11:11+10:00' // converts to the local browser timezone of `UTC+10`
```

I've opened a draft WC PR with a proposed fix: https://github.com/woocommerce/woocommerce/pull/53295.

**Comment by Jinksi** [comment]
After some trial and error, there is a workaround without modifying `@woocommerce/components` `Timeline` – subtract the browser timezone offset from the date values before passing to `Timeline`:

```js
date: new Date(
	moment
		.unix( event.datetime )
		// subtract the browser timezone offset to correct for the timezone difference
		.subtract( new Date().getTimezoneOffset(), 'minutes' )
		.valueOf()
),
```

Although, we should resolve this at the source (WC) if it is suitable.

**Comment by haszari** [comment]
Great investigation @Jinksi - this shows how tricky it can be dealing with timezones. For example, if we use the browser/user timezone for timeline (i.e. adopt the component's default), we may need to follow same approach elsewhere in WooPayments.

> If instead Timeline used @wordpress/date datei18n() to use the site/store's WP Settings timezone offset, the rendered datetimes would match the rendered time in the header.

So to be clear, we ideally need new behaviour in `Timeline` component? If so, perhaps we'd need to expose a parameter so clients of `Timeline` can opt-in to the new behaviour.

> there is a workaround without modifying @woocommerce/components Timeline – subtract the browser timezone offset from the date values before passing to Timeline:

This is a great example of the kind of workaround we want to avoid! 😁

**Comment by Jinksi** [comment]
> So to be clear, we ideally need new behaviour in `Timeline` component? If so, perhaps we'd need to expose a parameter so clients of `Timeline` can opt-in to the new behaviour.

Yep, assuming we need to change to store TZ, which is likely but not for certain. I've opened a WIP PR to expose a `Timeline` prop for either browser/store TZ use-case and confirm that `Timeline` is the source of the browser TZ conversion (TBC).

**Comment by Jinksi** [comment]
I've opened WC issue https://github.com/woocommerce/woocommerce/issues/53472 to represent this WC `Timeline` component issue/enhancement of requiring an option to render date times in the store's TZ rather than browser TZ.

**Comment by flakas** [comment]
Looks like this was fixed by https://github.com/Automattic/woocommerce-payments/pull/11895 (merged 2026-06-30). The timeline renders in the site's timezone now rather than the browser's, so it agrees with the Date shown above it.

Worth noting before this gets archived: the fix leans on the `timezone` prop added in https://github.com/woocommerce/woocommerce/pull/65914. On an older WooCommerce core that prop is ignored and the two can still disagree.

#9184 and #9472 are about the transactions list rather than this screen, so they stay open. Reopen this if you can still get a mismatch on the details screen with a current WC core.
